### ISSUE TITLE
pessimistic command pattern

### ISSUE THREAD

**Issue by ccutrer** [boc-tothefuture/openhab-jruby#275]
a very common pattern I see in my rules is `Item << something unless Item == something`. I.e. don't command a thing to do something if it's already in that state, because sending the command can be detrimental to the network (especially ZWave, when my items support instant status, so I trust OpenHAB to be up to date, and the network can get really slow with over a hundred nodes). 

My initial idea is to do
```
OptimisticWrapper = Struct.new(:item) do
  def <<(command)
    return if item.state == command
    item << command
  end
end

def optimistic(item)
  OptimisticWrapper.new(item)
end
```

so that I can then do `optimistic(MyDimmer) << 50`. But that doesn't work with helper-command-methods, like `optimistic(MyDimmer).on`. Simply delegating everything to the underlying item would lose the optimistic-ness of it. So what are thoughts on adding 

```
def optimistic
  @optimistic = true
  self
end

def command(command)
  return if @optimistic && state == command
  super
end
```

to GenericItem (or equivalent)? That way even the helper command methods would flow through the optimistic logic. And you could write `MyDimmer.optimistic.on`. One potentially confusing thing is how long that object is "optimistic" for. It obviously wouldn't leak into other rules files. It would stick if you've assigned the Item to a variable. But I'm not sure if it would for accessing items via const_missing, which may or may not be a good thing.

**Comment by boc-tothefuture** [comment]
Is it because you have multiple commands in the run block?  The rule block has support for guards already:

```
rule 'foo' do
run { item << something}
only_if { item != something }
end
```

If you are in the middle of a bunch of other commands in your run block it isn't as clean. 

Not against he idea of "command if different" in the style of "put if absent", but wanted to make sure you knew of existence of guards.

**Comment by ccutrer** [comment]
yup. doing lots of stuff like that. this is from a friend's config: https://github.com/ktgeek/openhab-stuff/blob/c373a63f2842acdb3379ceddcf1984125a1d75d0/config/automation/jsr223/ruby/personal/basement.rb#L10

I haven't converted as much of my stuff over to ruby yet, but https://github.com/ccutrer/openhab-config/blob/master/rules/switches.rules#L171 is an example in my own. I have lots of others, but I also have LOTS of rules to dig through to find them. (Sorry to general public... I don't publish my openhab config publicly for privacy reasons).

**Comment by boc-tothefuture** [comment]
Yeah, not optimal at all, very noisy for no benefit. 

In the current proposed method at the start of every rule you are going to have to set optimistic for each item you want to make optimistic. 

Couple of ideas just to think about and I haven't validated the technical feasibility of any of these:

First class support for optimistic rules in rule block, where all commands are treated as optimistic
```
rule 'foo' do
  run { item << ON}
  optimistic true
end
```

or

rule 'foo' do
  run (optimistic: true) do
     item << ON
     item2.off
  end
end
```

or

Maybe you specify the items you want to treat optimistically in the block? 
rule 'foo' do
  run do
     item.on
     item2.off
     item3 << 50
  end
optimistic item, item3
end
```

Another option would be a different operator for optimistic commands, but that would get confusing possibly.  I think it probably gets implemented the way you specified, but it is more about how you handle decoration of the object.

**Comment by jimtng** [comment]
I first raised this topic in #19. Cody's suggestion  Optimistic command should be done individually, not inside a block. Furthermore I don't think it should "stick" i.e.

when SwitchItem1.optimistic.on is followed by SwitchItem1.on, the second one shouldn't be optimistic. 

Also would someone who is unfamiliar with it, know immediately what 'optimistic' means?

**Comment by boc-tothefuture** [comment]
The only reason to tie it to the block is in most of the examples I have seen all of the elements in the block are treated the same way.  

It seems the goal here is to save keystrokes for a very common use case in openhab, if we are typing optimistic a bunch are we accomplishing that?  

Looking at the examples provided, they are a bit onerous.   But tying it to the entire runblock is maybe a bit limiting.  Perhaps we treat it the same way as the unit block?

So, you could have
```
SwitchItem1.optimistic.on

or
optimistic {
  SwitchItem.on
  SwitchItem2.off
}

```

**Comment by pacive** [comment]
Would 'ensure' even work, since it's a keyword? Other than that I don't have an opinion on this, I usually just send the command, regardless of what state the item's in. Haven't had any issues with that approach, but can understand that it might cause issues with larger ZigBee/z-wave networks.

**Comment by ccutrer** [comment]
`ensure` works for `SwitchItem.ensure.on`, but _not_ `ensure { SwitchItem.on }`. I can see myself using both inline form and block form. Would it be weird to use `optimistic` for block, but `ensure` for inline? another possibility: `only_if_not` ( which seems like it makes more sense in block form, but less so inline form)

**Comment by jimtng** [comment]
How would the syntax work for other commands e.g. NumberItem, String, etc. e.g. `NumberItem1 << 10`

**Comment by ccutrer** [comment]
Just the same. `NumberItem1.ensure << 10`

**Comment by boc-tothefuture** [comment]
Doh!  I forgot ensure was already a keyword with Rescue!

**Comment by boc-tothefuture** [comment]
`verify`?

```
verify { SwitchItem.on }
SwitchItem.verify.on
SwitchItem.verify << 10
```

**Comment by jimtng** [comment]
To me, `ensure` makes the most sense. How about `ensure_state` - I know it's longer.

**Comment by boc-tothefuture** [comment]
ensure_state seems to long for some cases.  I could get behind a hybrid:

```
ensure_state { SwitchItem.on }
SwitchItem.ensure.on
SwitchItem.ensure << 10
```

@ccutrer and @jimtng ?

**Comment by boc-tothefuture** [comment]
IMO, it belongs in the core.. It is highly useful and reusable. 

Like, I wouldn't but a PID algorithm in the core.. But this belongs.

**Comment by jimtng** [comment]
> ensure_state seems to long for some cases. I could get behind a hybrid:
> 
> ```
> ensure_state { SwitchItem.on }
> SwitchItem.ensure.on
> SwitchItem.ensure << 10
> ```
> 
> @ccutrer and @jimtng ?

Let's go with it! Maybe, alias ensure_state with ensure, so one can also do SwitchItem.ensure_state.on ? Either way is fine.

This will tidy up a lot of my code because I have a lot of `Item1.on unless Item1.on?` currently.
